src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -52,7 +52,36 @@ public static IntPtr Add<T>(this IntPtr start, int index)
         /// <summary>
         /// Determine if a type is eligible for storage in unmanaged memory. TODO: To be replaced by a ContainsReference() api.
         /// </summary>
-        public static bool IsReferenceFree<T>() => PerTypeLatches<T>.IsReferenceFree;
+        public static bool IsReferenceFree<T>()
+        {
+            // Under the JIT, these become constant-folded.
```

Comment:
What if those checks are moved to `IsReferenceFree(Type)` before the `if (type.GetTypeInfo().IsPrimitive) return true;` check?